app/src/main/java/de/westnordost/streetcomplete/quests/amenity_indoor/AddIsAmenityIndoor.kt

```diff
@@ -0,0 +1,114 @@
+import de.westnordost.streetcomplete.util.math.LatLonRaster
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+import de.westnordost.streetcomplete.util.math.isInMultipolygon
+import java.util.concurrent.FutureTask
+
+class AddIsAmenityIndoor(private val featureDictionaryFuture: FutureTask<FeatureDictionary>) :
+    OsmElementQuestType<Boolean> {
+
+    private val nodesFilter by lazy {
+        """
+        nodes with
+         (
+         emergency ~ defibrillator|fire_extinguisher
+         or amenity ~ atm|telephone|parcel_locker|luggage_locker|locker|clock|post_box|public_bookcase|give_box|ticket_validator|vending_machine
+         )
+         and access !~ private|no
+         and !indoor and !location and !level and !level:ref
+    """.toElementFilterExpression()
+    }
+
+    /* We only want survey nodes within building outlines. */
+    private val BuildingFilter by lazy {
+        """
+    ways, relations with building
+    """.toElementFilterExpression()
+    }
+
+    override val changesetComment = "Determine whether amenities are inside buildings"
+    override val wikiLink = "Key:indoor"
+
+    //TODO: Generisches Item
+    override val icon = R.drawable.ic_quest_defibrillator
+    override val achievements = listOf(CITIZEN)
+
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_is_amenity_inside_title
+
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+        val bbox = mapData.boundingBox ?: return listOf()
+        val nodes = mapData.nodes.filter { isApplicableTo(it) }
+        val buildings = mapData.filter { BuildingFilter.matches(it) }.toMutableList()
+
+        val buildingGeometriesById = buildings.associate {
+            it.id to mapData.getGeometry(it.type, it.id) as? ElementPolygonsGeometry
+        }
+
+        val nodesPositions = LatLonRaster(bbox, 0.0005)
+        for (node in nodes) {
+            nodesPositions.insert(node.position)
+        }
+
+        buildings.removeAll { building ->
+            val buildingBounds = buildingGeometriesById[building.id]?.getBounds()
+            (buildingBounds == null || !buildingBounds.isCompletelyInside(bbox) || nodesPositions.getAll(buildingBounds).count() == 0)
+        }
+
+        //Reduce all matching nodes to nodes within building outlines
+        val nodesInBuildings = nodes.filter {
+            buildings.any { building ->
+                val buildingGeometry = buildingGeometriesById[building.id]
+
+                if (buildingGeometry != null  && buildingGeometry.getBounds().contains(it.position) )
+                    it.position.isInMultipolygon(buildingGeometry.polygons)
+                 else
+                    false
+            }
+
+        }
+
+        return nodesInBuildings
+    }
+
+    override fun isApplicableTo(element: Element) =
+        nodesFilter.matches(element) && hasAnyName(element.tags)
```

Comment:
`isApplicableTo` is called also offline for any element that is updated due to the user changing it.

So for example when you solve the building quest, the app subsequently checks with this function whether for the new version of the building any new quests apply. In this case, the building levels quest will return true here, so a new quest is created.

Now, this means for your implementation here that if any e.g. `amenity=atm` is updated through any other quest (e.g. checking whether it is still there), the method here will always return true, ie.e. the `AddIsAmenityIndoor` quest is created. We do not want that.
In fact, as long as we do not have any information about the surrounding buildings, we cannot answer whether it applicable or not. Or well, if the given element does not match the `nodesFilter` or does not have any name, then we can say that it is not applicable, but otherwise, we can't tell. In this case, we want to return `null`.

Also have a loot at for example the housenumber quest as it also depends on the surrounding building geometry.

Reply:
Ok, i'm not sure if i understand this right.
Would it just be 
`override fun isApplicableTo(element: Element) {
if (!nodesFilter.matches(element) || !hasAnyName(element.tags)) false else null
}`

or should i think of a more complex filtering?

Reply:
For this function, yes. However, you use this function also in `getApplicableElements`, so there you should probably inline `nodesFilter.matches(element) && hasAnyName(element.tags)`